### PR TITLE
Homework 10 task 1

### PR DESCRIPTION
**Pull request by ivanzuev78**

### CODE REVIEW

File: HomeWork_10/task_threads.py [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+    return list(sorted(data, key=lambda x: float(x["1_year_growth"][:-1]))[-10:])
+
+
+def get_top_10_profit_52_weeks(data: List[Dict]) -> List[Dict]:
+    return list(sorted(data, key=lambda x: x["profit_52_weeks"])[-10:])
+
+
+def write_json_file(data: List[Dict], filename: str) -> None:
+    with open(filename, "w") as file:
+        json.dump(data, file)
```
Comment (Cjkjvfnby) [inline HomeWork_10/task_threads.py:121]: I recommend adding indent=4 for readability.

File: HomeWork_10/task_threads.py [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+        all_html_main_pages_with_companies = pool.map(
+            requests.get, all_urls_main_pages_with_companies
+        )
+        all_companies_dicts_before_parsing_its_page = list(
+            pool.map(get_all_companies_from_page, all_html_main_pages_with_companies)
+        )
+        all_companies_dicts_after_parsing_its_page = list(
+            pool.map(
+                get_info_from_company_url,
+                itertools.chain(*all_companies_dicts_before_parsing_its_page),
```
Comment (Cjkjvfnby) [inline HomeWork_10/task_threads.py:140]: What argument type is expected here?

File: HomeWork_10/task_threads.py [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+    for i in bs_obj.find("div", class_="finando_paging").find_all("a"):
+        paginator_pages.append(
+            "https://markets.businessinsider.com/index/components/s&p_500" + i["href"]
+        )
+    return paginator_pages
+
+
+def get_all_companies_from_page(html: Response) -> List[Dict]:
+    companies = []
+    bsObj = BeautifulSoup(html.text, "html.parser")
```
Comment (Cjkjvfnby) [inline HomeWork_10/task_threads.py:25]: This name looks as strange as https://ru.wikipedia.org/wiki/%D0%91%D0%9E%D0%A7_%D1%80%D0%92%D0%A4_260602

You can name it `content`.

File: HomeWork_10/task_threads.py [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+    for company in bsObj.find("main", class_="site-content").find_all(
+        "a", {"href": re.compile("/stocks/\w*")}
+    ):
+        companies.append(
+            {
+                "company_url": "https://markets.businessinsider.com" + company["href"],
+                "name": company["title"],
+            }
+        )
+    for j, comp in zip(bsObj.find_all("img"), companies):
```
Comment (Cjkjvfnby) [inline HomeWork_10/task_threads.py:35]: j stands for jif?  https://time.com/5791028/how-to-pronounce-gif/

File: HomeWork_10/task_threads.py [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+        if re.match("https://pproxy.markets.businessinsider.com", j["src"]) != "None":
```
Comment (Cjkjvfnby) [inline HomeWork_10/task_threads.py:36]: `!= "None"`  Does re.match return a string?

File: HomeWork_10/task_threads.py [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+
+def convert_usd_to_rub_in_company(company: Dict, usd_to_rub: float) -> Dict:
+    company["price_in_RUB"] = company["price_in_USD"] * usd_to_rub
+    return company
+
+
+def get_top_10_most_expensive_companies(
+    data: List[Dict], usd_to_rub: float
+) -> List[Dict]:
+    top_10 = list(sorted(data, key=lambda x: x["price_in_USD"])[-10:])
```
Comment (Cjkjvfnby) [inline HomeWork_10/task_threads.py:93]: What is the return type of `sorted`?

Comment (Cjkjvfnby) [reply HomeWork_10/task_threads.py:93]: https://docs.python.org/3/library/heapq.html#heapq.nlargest

File: HomeWork_10/task_threads.py [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+    for company in top_10:
+        company["price_in_RUB"] = company["price_in_USD"] * usd_to_rub
```
Comment (Cjkjvfnby) [inline HomeWork_10/task_threads.py:95]: What is bad with this code (expression itself is ok)?

File: HomeWork_10/task_threads.py [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+    return top_10
+
+
+def get_top_10_min_pe(data: List[Dict]) -> List[Dict]:
```
Comment (Cjkjvfnby) [inline HomeWork_10/task_threads.py:99]: Complexity overflow in a single expression. 

Please split it into chunks by extracting values to the variables.

File: HomeWork_10/task_threads.py [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+    return list(
+        filter(
+            lambda x: True if "P/E" in x else False,
```
Comment (Cjkjvfnby) [inline HomeWork_10/task_threads.py:102]: Can you simplify the expression?

Which module can give us simple replacement for such types of lambda?

File: HomeWork_10/test_HW_10_task_1.py [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+    get_paginator_pages_sp500,
+    get_top_10_min_pe,
+    get_top_10_most_expensive_companies,
+    get_top_10_profit_52_weeks,
+    get_top_10_year_growth,
+)
+
+import pytest
+
+os.chdir(os.path.dirname(__file__))
```
Comment (Cjkjvfnby) [inline HomeWork_10/test_HW_10_task_1.py:16]: I'd prefer moving this to a fixture.  
You can set the scope module and set autouse, or do it in fixtures that reads the file.

**Review by Cjkjvfnby — CHANGES_REQUESTED** [review]

File: HomeWork_10/task_threads.py (hunk 1 above)
Comment (ivanzuev78) [reply HomeWork_10/task_threads.py:121]: I don't understand, what you mean.

File: HomeWork_10/task_threads.py (hunk 2 above)
Comment (ivanzuev78) [reply HomeWork_10/task_threads.py:140]: Iterable is expected. 
all_companies_dicts_before_parsing_its_page is List[List[Dict]].
But we need List[Dict]. So, with itertools.chain we unpack inner lists.

File: HomeWork_10/task_threads.py (hunk 3 above)
Comment (ivanzuev78) [reply HomeWork_10/task_threads.py:25]: You right!

File: HomeWork_10/task_threads.py (hunk 4 above)
Comment (ivanzuev78) [reply HomeWork_10/task_threads.py:35]: J is jpg. I wanted to change it later but had forgotten.

File: HomeWork_10/task_threads.py (hunk 5 above)
Comment (ivanzuev78) [reply HomeWork_10/task_threads.py:36]: No.

File: HomeWork_10/task_threads.py (hunk 6 above)
Comment (ivanzuev78) [reply HomeWork_10/task_threads.py:93]: Return type of `sorted` is an iterator. Using `list` was unnecessary. 
I don't know about `nlargest`.
Thanks.

File: HomeWork_10/task_threads.py (hunk 7 above)
Comment (ivanzuev78) [reply HomeWork_10/task_threads.py:95]: I need one more hint.

Comment (ivanzuev78) [reply HomeWork_10/task_threads.py:95]: Maybe we should take `nlargest` by price in rub?

File: HomeWork_10/task_threads.py (hunk 8 above)
Comment (ivanzuev78) [reply HomeWork_10/task_threads.py:99]: You meant something like these: `def get_top_10_min_pe(*data: Dict)` ?

File: HomeWork_10/task_threads.py (hunk 9 above)
Comment (ivanzuev78) [reply HomeWork_10/task_threads.py:102]: Expression simplified.
I don't remember the module, but I remember, how I can simplify this lambda.

File: HomeWork_10/test_HW_10_task_1.py (hunk 10 above)
Comment (ivanzuev78) [reply HomeWork_10/test_HW_10_task_1.py:16]: Done.

File: HomeWork_10/task_threads.py (hunk 1 above)
Comment (Cjkjvfnby) [reply HomeWork_10/task_threads.py:122]: https://docs.python.org/3/library/json.html#json.dump

File: HomeWork_10/task_threads.py (hunk 7 above)
Comment (Cjkjvfnby) [reply HomeWork_10/task_threads.py:98]: It is duplicated.

**Review by Cjkjvfnby — CHANGES_REQUESTED** [review]
Some small fixes required.

See https://docs.python.org/3/library/operator.html?highlight=operator#operator.contains

File: HomeWork_10/task_threads.py (hunk 1 above)
Comment (ivanzuev78) [reply HomeWork_10/task_threads.py:122]: If I only know about this, my life would be much easier.

File: HomeWork_10/task_threads.py (hunk 2 above)
Comment (ivanzuev78) [reply HomeWork_10/task_threads.py:141]: Seems good.

Comment (ivanzuev78) [reply HomeWork_10/task_threads.py:141]: But `pool.map` trying to send values from all iterators to the function.
TypeError: get_info_from_company_url() takes 1 positional argument but 10 were given
I tried:
1) pool.map(get_info_from_company_url, all_companies_dicts_before_parsing_its_page)
2) pool.map(get_info_from_company_url, *all_companies_dicts_before_parsing_its_page)

File: HomeWork_10/task_threads.py (hunk 7 above)
Comment (ivanzuev78) [reply HomeWork_10/task_threads.py:98]: Hmm, it is not obvious to me.

Comment (ivanzuev78) [reply HomeWork_10/task_threads.py:90]: Oh, I am inattentive a little bit.

**Comment by ivanzuev78** [comment]
Unused function deleted.

**Review by Cjkjvfnby — APPROVED** [review]

**Pull request merged**